src/insexpand.c

```diff
@@ -1258,8 +1269,76 @@ ins_compl_build_pum(void)
 	    if (match_head == NULL)
 		match_head = compl;
 	    else
-		match_tail->cp_match_next  = compl;
+		match_tail->cp_match_next = compl;
 	    match_tail = compl;
+
+	    if (!shown_match_ok && !compl_fuzzy_match)
+	    {
+		if (compl == compl_shown_match || did_find_shown_match)
+		{
+		    // This item is the shown match or this is the
+		    // first displayed item after the shown match.
+		    compl_shown_match = compl;
+		    did_find_shown_match = TRUE;
+		    shown_match_ok = TRUE;
+		}
+		else
+		    // Remember this displayed match for when the
+		    // shown match is just below it.
+		    shown_compl = compl;
+		cur = i;
+	    }
+	    else if (compl_fuzzy_match)
+	    {
+		if (i == 0)
+		    shown_compl = compl;
+		// Update the maximum fuzzy score and the shown match
+		// if the current item's score is higher
+		if (compl->cp_score > max_fuzzy_score)
+		{
+		    did_find_shown_match = TRUE;
+		    max_fuzzy_score = compl->cp_score;
+		    if (!compl_no_select)
+			compl_shown_match = compl;
+		}
+
+		if (!shown_match_ok && compl == compl_shown_match && !compl_no_select)
+		{
+		    cur = i;
+		    shown_match_ok = TRUE;
+		}
+
+		// If there is no "no select" condition and the max fuzzy
+		// score is positive, or there is no completion leader or the
+		// leader length is zero, mark the shown match as valid and
+		// reset the current index.
+		if (!compl_no_select
+			&& (max_fuzzy_score > 0
+				|| (compl_leader.string == NULL || compl_leader.length == 0)))
```

Comment:
```suggestion
				|| compl_leader.length == 0))
```
I don't think compl_leader.length can be positive while compl_leader.string is NULL.

Reply:
leader can be an empty string the length is 0 but is not NULL